core/error.rs

```diff
@@ -1,143 +1,260 @@
+  #[error("Failed to parse {0}")]
+  Parse(FastStaticString),
+  #[error("Failed to execute {0}")]
+  Execute(FastStaticString),
+  #[error(
+    "Following modules were passed to ExtModuleLoader but never used:\n{}",
+    .0.iter().map(|s| format!("  - {}\n", s)).collect::<Vec<_>>().join("")
+  )]
+  UnusedModules(Vec<String>),
+  #[error(
+    "Following modules were not evaluated; make sure they are imported from other code:\n{}",
+    .0.iter().map(|s| format!("  - {}\n", s)).collect::<Vec<_>>().join("")
+  )]
+  NonEvaluatedModules(Vec<String>),
+  #[error("{0} not present in the module map")]
+  MissingFromModuleMap(String),
+  #[error(transparent)]
+  ModuleLoader(Box<ModuleLoaderError>),
+  #[error("Could not execute {specifier}")]
+  CouldNotExecute {
+    #[source]
+    error: Box<Self>,
+    specifier: String,
+  },
+  #[error(transparent)]
+  JsNative(#[from] deno_error::JsErrorBox),
+  #[error(transparent)]
+  Url(#[from] url::ParseError),
+  #[error(transparent)]
+  FutureCanceled(#[from] futures::channel::oneshot::Canceled),
+  #[error(
+    "Cannot evaluate module, because JavaScript execution has been terminated"
+  )]
+  ExecutionTerminated,
+  #[error("Promise resolution is still pending but the event loop has already resolved"
+  )]
+  PendingPromiseResolution,
+  #[error("Cannot evaluate dynamically imported module, because JavaScript execution has been terminated"
+  )]
+  EvaluateDynamicImportedModule,
+  #[error(transparent)]
+  Module(ModuleConcreteError),
+  #[error(transparent)]
+  DataError(DataError),
+  #[error(transparent)]
+  Other(#[from] anyhow::Error),
 }
 
-pub fn generic_error(message: impl Into<Cow<'static, str>>) -> Error {
-  custom_error("Error", message)
-}
+impl CoreError {
+  pub fn print_with_cause(&self) -> String {
+    use std::error::Error;
+    let mut err_message = self.to_string();
 
-pub fn type_error(message: impl Into<Cow<'static, str>>) -> Error {
-  custom_error("TypeError", message)
-}
-
-pub fn range_error(message: impl Into<Cow<'static, str>>) -> Error {
-  custom_error("RangeError", message)
-}
+    if let Some(source) = self.source() {
+      err_message.push_str(&format!(
+        "\n\nCaused by:\n    {}",
+        source.to_string().replace("\n", "\n    ")
+      ));
+    }
 
-pub fn invalid_hostname(hostname: &str) -> Error {
-  type_error(format!("Invalid hostname: '{hostname}'"))
-}
+    err_message
+  }
 
-pub fn uri_error(message: impl Into<Cow<'static, str>>) -> Error {
-  custom_error("URIError", message)
-}
+  pub fn to_v8_error(
+    &self,
+    scope: &mut v8::HandleScope,
+  ) -> v8::Global<v8::Value> {
+    let err_string = self.get_message().to_string();
+    let mut error_chain = vec![];
+    let mut intermediary_error: Option<&(dyn Error)> = Some(&self);
+
+    while let Some(err) = intermediary_error {
+      if let Some(source) = err.source() {
+        let source_str = source.to_string();
+        if source_str != err_string {
```

Comment:
Is this check necessary?

Reply:
i believe we were doing similar handling like this somewhere before, and when i tried removing it, some stuff broke, so didnt tihnk too much about it and just kept it as it is

Reply:
oh interesting, didn't know we were already doing something like this. seems fine then

Reply:
or i might be going insane